File: compiler/server/ide_info.go
package server

import "encoding/json"

type IdeInfoRequest struct {
	InputMetadata
}

type IdeInfoResponse struct {
	Info []json.RawMessage `json:"info"`
}

func (request *IdeInfoRequest) handle() (IdeInfoResponse, error) {
	var info []json.RawMessage

	libraryName := request.Library
	for libraryName != "" {
		library, _, err := fetchLibrary(libraryName)
		if err != nil {
			return IdeInfoResponse{}, err
		}

		if library.Metadata.Ide != nil {
			info = append(info, library.Metadata.Ide)
		}

		libraryName = library.Metadata.Library
	}

	return IdeInfoResponse{Info: info}, nil
}

File: compiler/server/ide_info.rs
use serde::Deserialize;
use serde_json::json;

use crate::{InputMetadata, libraries::fetch_library};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Request {
    #[serde(flatten)]
    metadata: InputMetadata,
}

pub async fn handle(request: Request) -> anyhow::Result<serde_json::Value> {
    let mut info = Vec::new();
    let mut current = request.metadata.library;
    while let Some(library) = current {
        let library = fetch_library(&library).await?;

        if let Some(item) = library.metadata.ide {
            info.push(item);
        }

        current = library.metadata.library;
    }

    Ok(json!({ "info": info }))
}
